xayn-ai/src/coi/system.rs

```diff
@@ -382,6 +376,48 @@ impl CoiSystem {
             .collect();
         coi.swap_key_phrases(key_phrases);
     }
+
+    /// Selects the top key phrases from the cois.
+    #[allow(dead_code)]
+    fn select_top_key_phrases<CP: CoiPoint + CoiPointKeyPhrases + CoiPointStats>(
+        &self,
+        cois: &mut [CP],
+        top: usize,
+    ) -> Vec<String> {
+        if top == 0 {
+            return Vec::new();
+        }
+
+        let relevances = self.compute_relevances(cois);
+        let mut relevant_key_phrases = cois
+            .iter_mut()
+            .zip(relevances)
+            .map(|(coi, relevance)| {
+                self.select_key_phrases(coi, &[]);
+                let mut key_phrases = coi
+                    .key_phrases()
+                    .iter()
+                    .map(|key_phrase| (key_phrase.words(), key_phrase.relevance()))
+                    .collect::<Vec<_>>();
+                key_phrases.sort_unstable_by(|(_, this), (_, other)| {
+                    this.partial_cmp(other).unwrap(/* relevance is never nan */)
+                });
+                key_phrases.into_iter().rev().zip(self.config.penalty).map(
+                    move |((key_phrase, _), penalty)| {
+                        (key_phrase, (relevance * penalty).max(f32::MIN))
```

Comment:
Why with need the max with min?

Reply:
to guarantee that the resulting value is clamped into finite floats (`f32::clamp()` won't do this).